src/gpb_compile.erl

```diff
@@ -6318,7 +6331,11 @@ can_specify_map_item_presence_in_typespecs(Opts) ->
 
 type_to_typestr(#?gpb_field{type=Type, occurrence=Occurrence}, Defs, Opts) ->
     OrUndefined = case get_mapping_and_unset_by_opts(Opts) of
-                      records                   -> " | undefined";
+                      records                   ->
+                          case proplists:get_value(syntax, Defs, yes) of
+                              "proto3" -> "";
+                              _ -> " | undefined"
+                          end;
```

Comment:
ditto here, if you pass it in. It will be available at the call sites once you've passed it in to the location above